website_cms/models/website.py

```diff
@@ -0,0 +1,244 @@
+        result = []
+        if pages is None:
+            search_args = [
+                ('parent_id', '=', False),
+                ('type_id', 'in', type_ids)
+            ]
+            if published is not None:
+                search_args.append(('website_published', '=', published))
+
+            if nav is not None:
+                search_args.append(('nav_include', '=', nav))
+
+            sec_model = self.env['cms.page']
+            pages = sec_model.search(
+                search_args,
+                order='sequence asc'
+            )
+        for item in pages:
+            result.append(
+                self._build_page_item(item,
+                                      max_depth=max_depth,
+                                      nav=nav,
+                                      type_ids=type_ids,
+                                      published=published)
+            )
+        return result
+
+    @api.model
+    def _build_page_item(self, item, max_depth=3,
+                         nav=True, type_ids=None,
+                         published=True):
+        """Recursive method to build main menu items.
+
+        Return a dict-like object containing:
+        * `name`: name of the page
+        * `url`: public url of the page
+        * `children`: list of children pages
+        * `nav`: nav_include filtering
+        * `type_ids`: filter pages by a list of types' ids
+        * `published`: filter pages by a publishing state
+        """
+        depth = max_depth or 3  # safe default to avoid infinite recursion
+        sec_model = self.env['cms.page']
+        # XXX: consider to define these args in the main method
+        search_args = [
+            ('parent_id', '=', item.id),
+        ]
+        if published is not None:
+            search_args.append(('website_published', '=', published))
+        if nav is not None:
+            search_args.append(('nav_include', '=', nav))
+        if type_ids is not None:
+            search_args.append(('type_id', 'in', type_ids))
+        subs = sec_model.search(
+            search_args,
+            order='sequence asc'
+        )
+        while depth != 0:
+            depth -= 1
+            children = [self._build_page_item(x, max_depth=depth)
+                        for x in subs]
+        res = AttrDict({
+            'id': item.id,
+            'name': item.name,
+            'url': item.website_url,
+            'children': children,
+            'website_published': item.website_published,
+        })
+        return res
+
+    @api.model
+    def safe_image_url(self, record, field, size=None, check=True):
+        """Return image url if exists."""
+        sudo_record = record.sudo()
+        if hasattr(sudo_record, field):
+            if not getattr(sudo_record, field) and check:
+                # no image here yet
+                return ''
+            return self.image_url(record, field, size=size)
+        return ''
```

Comment:
Why sudo?

This could be:
```
try:
    content = record.field
except (AttributeError, AccessError):
    return ""
return self.image_url(record, field, size=size) if content or not check else ""
```